### PR TITLE
Add part 3 md file and visualizations 

### PR DESCRIPTION
**Pull request by fox-daniel**
- Add md file for part 3 to src.
- Add .html files with visualizations for part 3 to src.
- Revise SUMMARY to include part 3 as a chapter.

### CODE REVIEW

File: src/iterative_methods_part_3.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+## Outline of the Post
+- The UI for Reservoir Sampling
+- Example: Visualizing the Evolving Distribution  
+- Example: Reservoir vs. Stream Means 
+- Example: Exporting Data for the Visualizations Using Adaptors
+
+## The UI for Reservoir Sampling  
+
+The UI uses adaptors to transform the behavior of `StreamingIterator`s. Suppose that `stream` is a `StreamingIterator` with items of type `T`. We adapt that iterator using `reservoir_iterable(stream, capacity, rng)`, whose fields are 1) a streaming iterator, 2) the capacity or size of the reservoir sample, and 3) a choice of a random number generator. The default is to set `rng` to `None`so that [`rand_pcg::Pcg64`](https://rust-random.github.io/book/intro.html) is used. You also have the option of using a seedable `rng` that might be useful for debugging or testing. Each item of the returned `StreamingIterator` is a `Vec<T>`, where the vector holds a reservoir sample.
```
Comment (daniel-vainsencher) [inline src/iterative_methods_part_3.md:17]: Here you call it "capacity or size", but before you called it "sample of k items". Might want to have some overlap (e.g., "capacity or sample size k") to help link those securely.

File: src/iterative_methods_part_3.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+This post describes how the `Iterative Methods` crate facilitates easy reservoir sampling of a `StreamingIterator`. [Reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling) produces an up-to-date and relatively low cost random sample of a large stream of data. For example, suppose you want to maintain an up-to-date sample of \\(k\\) tweets from a twitter feed. At any moment, a reservoir sample of the tweets is equivalent to a random sample of \\(k\\) items from the portion of the stream that has been processed at that moment. The reservoir sampling algorithm accomplishes this without needing to know the total number of samples in the stream and it updates the sample to take into account new behavior in the data that may not have been present initially. Below we'll see how to use reservoir sampling for `StreamingIterator`s in Rust. Using animations we can see how the reservoir samples stay up-to-date as the data stream exhibits new behavior. 
+
+## Outline of the Post
+- The UI for Reservoir Sampling
+- Example: Visualizing the Evolving Distribution  
+- Example: Reservoir vs. Stream Means 
+- Example: Exporting Data for the Visualizations Using Adaptors
```
Comment (daniel-vainsencher) [inline src/iterative_methods_part_3.md:13]: I don't have a good suggestion, but I feel like this TOC is not so strong. A silly reader might think you present one idea (ResSampl) and then just 3 redundant examples of it?

File: src/iterative_methods_part_3.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+</figure>
+
+## Example: Reservoir vs. Stream Means
+
+The animation in Figure 2 allows us to see how the reservoir distribution tracks the stream distribution. We can also check that the mean of the reservoir is close to the mean of the portion of the stream that has been processed. Now, if you want to compute streaming means, what follows is not the efficient way to do it. Rather, what follows is useful for checking that the means of the reservoirs are behaving as expected. The `Iterative Methods` crate allows us to use the kind of flexible adaptors you are used to from Rust's `Iterator`s to accomplish this. We'll plot the reservoir mean vs. the mean of the portion of the stream that was sampled to produce the reservoir.
```
Comment (daniel-vainsencher) [inline src/iterative_methods_part_3.md:50]: two consecutive uses of "what follows", and the paragraph is a bit long and awkward. Seems to me the 2nd, 4th and 6th sentences are closely related, perhaps should come together? 5th seems like a good segue to next paragraph.

File: src/iterative_methods_part_3.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+Here is the code that accomplishes this. The code is modular; once the data stream exists we adapt, adapt, adapt in whichever sequence is currently useful. Again, we suppose that `stream` is our `StreamingIterator` full of float samples from the pair of distributions as described above. The `stream` starts off with each item an `f64`; 
+- After the `enumerate` adaptor, the items are `Numbered<f64>` with indices that will allow us to calculate `max_index` for each reservoir;
+- After the `reservoir_iterable()` each item is a reservoir sample in a `Vec<Numbered<f64>>`;
+- The `map` adaptor uses the named closure `reservoir_mean_and_max_index` to compute the mean and maximum index for each reservoir.  After the `map` adaptor each item is a `Numbered` struct containing the mean of the reservoir and the `max_index` indicating how much of the stream was sampled to obtain that reservoir. See the source code for the wheels within wheels.
```
Comment (daniel-vainsencher) [inline src/iterative_methods_part_3.md:57]: wheels within wheels... is that a Dune reference? by referring to source code, do you mean below, or the repo? if the repo, perhaps permalink to a particular file?

File: src/iterative_methods_part_3.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+- - 2
+  - 0.29291775753278054
+- - 3
+  - 0.5312252557573507
+```
+<figcaption style="text-align:center;">Code Block 4. The head of the YAML file containing the reservoir samples. The items in the reservoir are (index, value) pairs. </figcaption>
+
+The visualizations were generated from the data in the YAML files using the [Plotly](https://plotly.com/python/) module in Python. In the future we hope to switch to an entirely Rusty solution using the [Plotters](https://docs.rs/plotters/0.3.0/plotters/) crate. Currently, our Rust program runs Python scripts using `std::process::Command` and writes errors they might throw to `stderr`.
+
+The iterator-adaptor idiom popular in Rust and other modern languages provides an ergonomic way to write code. As we build a repertoire of adaptors that implement useful iterative methods, we can easily deploy them in myriad combinations to meet our engineering needs. For example, the YAML adaptor can also be used for checkpoints, logs, and other streaming needs. If you try out the `Iterative Methods` crate, please send us feedback!
```
Comment (daniel-vainsencher) [inline src/iterative_methods_part_3.md:143]: Love the call to action in the end, perhaps invite the reader to run the particular cargo run --example invocation?

**Review by daniel-vainsencher — APPROVED** [review]
A few minor suggestions; implement those you like and land. I'll add a commit with a link to it from the preceding blog post.

**Comment by daniel-vainsencher** [comment]
Oh and also please run mdbook before pushing and landing, that's what renders the blog itself.

**Comment by daniel-vainsencher** [comment]
One more thing: I've renamed reservoir_iterable to reservoir_sample (and similar the struct) and that affects your post :/ something like `sed 's/reservoir_iterable/reservoir_sample/' -i src/*.md` might do the trick, but do check.

File: src/iterative_methods_part_3.md (hunk 1 above)
Comment (fox-daniel) [reply src/iterative_methods_part_3.md:17]: good point.

File: src/iterative_methods_part_3.md (hunk 3 above)
Comment (fox-daniel) [reply src/iterative_methods_part_3.md:50]: Good point. Good point. I reordered and reworded.

File: src/iterative_methods_part_3.md (hunk 4 above)
Comment (fox-daniel) [reply src/iterative_methods_part_3.md:57]: I linked to the examples folder. The actual file is now mid rename as I opened a PR to rename it and do some other cleanup.

**Pull request merged**